File: linked_list/cpp/remove-node.cpp
#include <iostream>
using namespace std;

class Node{
  public:
    Node* next;
    int value;
    Node(int v) { this->value = v; }
};

/**
 Remove all the node whose value=v in the given list, 
 return the new head of the list.
 e.g. 
 input list:  3 -> 4 -> 8 -> 6 -> 4, and number 4,
 output list: 3 -> 8 -> 6, and return 2
 test case:
   linked list(value to delete)
   NULL(4)
   4(4)
   4(3)
   1 -> 3 -> 2(1)
   4 -> 4 -> 4(4)
   1 -> 3 -> 2(2)
 */
Node* removeNode(Node* head, int v){
  if (head == NULL) return 0;
  Node* pre = NULL;
  Node* cur = head;
  int count = 0;
  while (cur != NULL) {
    if (cur->value == v) {
      // remove this node
      if(pre != NULL) pre->next = cur->next; // remove non-head not
      else head = cur->next; // remove head node
      delete cur;
      count++;
      cur = pre == NULL ? head : pre->next;
    } else {
      // moving forword
      pre = cur;
      cur = cur->next;
    } 
  }
  cout << "Number of nodes deleted: " << count << endl;
  return head;
}

Node* createLinkedList(int* arr, int length){
  if (arr == NULL) return NULL;
  Node* pre = new Node(arr[0]);
  Node* head = pre;
  for (int i=1; i<length; i++){
    Node* n = new Node(arr[i]);
    pre->next = n;
    pre = n;
  }
  return head;
}

void printLinkedList(Node* head) {
  if (head == NULL) {
    cout << "empty list" << endl;
    return;
  }
  Node* p = head;
  while(p != NULL) {
    cout << p->value << " -> ";
    p = p->next;
  }
  cout << "NULL" << endl;
}

int main(){
  int values[] = {4, 4, 4, 1, 2, 5};
  int length = sizeof(values) / sizeof(int);
  Node* l1 = createLinkedList(values, length);
  printLinkedList(l1);
  Node* newHead = removeNode(l1, 4);
  printLinkedList(newHead);
  return 0;
}

File: linked_list/java/RemoveNode.java
//package linkedlist;
//import linkedlist.LinkedListHelper;

public class RemoveNode {
    public static void main(String[] args) {
        int[] arr = {4, 4, 4, 8, 7, 6};
        Node head = LinkedListHelper.createLinkedList(arr);
        LinkedListHelper.printLinkedList(head);
        Node removedHead = removeNode(head, 8);
        LinkedListHelper.printLinkedList(head);
        removedHead = removeNode(removedHead, 4);
        LinkedListHelper.printLinkedList(removedHead);
    }

    // remove all nodes whose value = v
    // return the head node after removal
    // test case 
    /*
        4, 8, 7, 6 => remove 8
        4, 8, 7, 6 => remove 4
        4, 4, 4    => remove 4
     */
    public static Node removeNode(Node head, int v) {
        // do some check
        if (head == null) return null;
        Node prev = null;
        Node cur = head;
        while(cur != null) {
            if (cur.value == v) {
                // we should remove cur
                if (prev == null) {
                    // we are removing the head node
                    head = head.next;
                    cur = head;
                } else {
                    // we are in the middle of the list
                    prev.next = cur.next;
                    cur.next = null;
                    cur = prev.next;
                }
            } else {
                // move cur and prev forward
                prev = cur;
                cur = cur.next;
            }
        }
        return head;
    }


}

